ql/src/test/queries/clientpositive/jdbc_sort_limit.q

```diff
@@ -0,0 +1,132 @@
+    "hive.sql.dbcp.username" = "qtestuser",
+    "hive.sql.dbcp.password" = "qtestpassword",
+    "hive.sql.table" = "date_dim"
+);
+
+CREATE EXTERNAL TABLE IF NOT EXISTS `store_sales`(
+  `ss_sold_date_sk` int,
+  `ss_sold_time_sk` int, 
+  `ss_item_sk` bigint, 
+  `ss_customer_sk` int, 
+  `ss_cdemo_sk` int, 
+  `ss_hdemo_sk` int, 
+  `ss_addr_sk` int, 
+  `ss_store_sk` int, 
+  `ss_promo_sk` int, 
+  `ss_ticket_number` bigint, 
+  `ss_quantity` int, 
+  `ss_wholesale_cost` decimal(7,2), 
+  `ss_list_price` decimal(7,2), 
+  `ss_sales_price` decimal(7,2), 
+  `ss_ext_discount_amt` decimal(7,2), 
+  `ss_ext_sales_price` decimal(7,2), 
+  `ss_ext_wholesale_cost` decimal(7,2), 
+  `ss_ext_list_price` decimal(7,2), 
+  `ss_ext_tax` decimal(7,2), 
+  `ss_coupon_amt` decimal(7,2), 
+  `ss_net_paid` decimal(7,2), 
+  `ss_net_paid_inc_tax` decimal(7,2), 
+  `ss_net_profit` decimal(7,2))
+STORED BY                                          
+'org.apache.hive.storage.jdbc.JdbcStorageHandler'
+TBLPROPERTIES (                                    
+    "hive.sql.database.type" = "POSTGRES",
+    "hive.sql.jdbc.driver" = "org.postgresql.Driver",
+    "hive.sql.jdbc.url" = "jdbc:postgresql://localhost:5432/qtestDB",
+    "hive.sql.dbcp.username" = "qtestuser",
+    "hive.sql.dbcp.password" = "qtestpassword",
+    "hive.sql.table" = "store_sales"
+);
+
+CREATE EXTERNAL TABLE IF NOT EXISTS `item`(
+  `i_item_sk` bigint, 
+  `i_item_id` string, 
+  `i_rec_start_date` string, 
+  `i_rec_end_date` string, 
+  `i_item_desc` string, 
+  `i_current_price` decimal(7,2), 
+  `i_wholesale_cost` decimal(7,2), 
+  `i_brand_id` int, 
+  `i_brand` string, 
+  `i_class_id` int, 
+  `i_class` string, 
+  `i_category_id` int, 
+  `i_category` string, 
+  `i_manufact_id` int, 
+  `i_manufact` string, 
+  `i_size` string, 
+  `i_formulation` string, 
+  `i_color` string, 
+  `i_units` string, 
+  `i_container` string, 
+  `i_manager_id` int, 
+  `i_product_name` string)
+STORED BY                                          
+'org.apache.hive.storage.jdbc.JdbcStorageHandler'
+TBLPROPERTIES (                                    
+    "hive.sql.database.type" = "POSTGRES",
+    "hive.sql.jdbc.driver" = "org.postgresql.Driver",
+    "hive.sql.jdbc.url" = "jdbc:postgresql://localhost:5432/qtestDB",
+    "hive.sql.dbcp.username" = "qtestuser",
+    "hive.sql.dbcp.password" = "qtestpassword",
+    "hive.sql.table" = "item"
+);
+
+set hive.mapred.mode=nonstrict;
+
+explain cbo
+select  dt.d_year 
+       ,item.i_brand_id brand_id 
+       ,item.i_brand brand
+       ,sum(ss_ext_sales_price) sum_agg
+ from  date_dim dt 
+      ,store_sales
+      ,item
+ where dt.d_date_sk = store_sales.ss_sold_date_sk
+   and store_sales.ss_item_sk = item.i_item_sk
+   and item.i_manufact_id = 436
+   and dt.d_moy=12
+ group by dt.d_year
+      ,item.i_brand
+      ,item.i_brand_id
+ order by dt.d_year
+         ,sum_agg desc
+         ,brand_id
+limit 100;
```

Comment:
The repro test is unnecessarily verbose making hard to understand where exactly is the problem and why it is fixed by the proposed changes. According to the description the problem should be visible in any query containing a limit.

How about just adding the following "jdbc_table_limit_postgres.q":
```sql
--! qt:database:postgres:q_test_country_table.sql

CREATE EXTERNAL TABLE country (id int, name varchar(20))
STORED BY 'org.apache.hive.storage.jdbc.JdbcStorageHandler'
TBLPROPERTIES (
    "hive.sql.database.type" = "POSTGRES",
    "hive.sql.jdbc.driver" = "org.postgresql.Driver",
    "hive.sql.jdbc.url" = "jdbc:postgresql://localhost:5432/qtestDB",
    "hive.sql.dbcp.username" = "qtestuser",
    "hive.sql.dbcp.password" = "qtestpassword",
    "hive.sql.table" = "country");

EXPLAIN CBO SELECT * FROM country ORDER BY id LIMIT 2;
EXPLAIN SELECT * FROM country ORDER BY id LIMIT 2;
SELECT * FROM country ORDER BY id LIMIT 2;
```
The above test case is minimal, and allows us to test all aspects of the plan (compilation+execution).

It may also be nice to reword the JIRA case using the minimal example.

Reply:
Thanks! I have added the test as suggested here.